### PR TITLE
Do not panic on string conversions in yaml unmarshaling.

### PR DESCRIPTION
**Pull request by imikushin**
Do not panic on string conversions in yaml unmarshaling.

Signed-off-by: Ivan Mikushin <i.mikushin@gmail.com>

### CODE REVIEW

**Comment by vdemeester** [comment]
Sounds fair :wink:
Does the test *testing* that it does not panic ? :P

**Comment by dnephin** [comment]
If I understand correctly, this will treat `true` as `'true'`, which I don't think is correct.

Does it also cast all numbers to strings?

Wouldn't this be better handled by #99 ?

**Comment by imikushin** [comment]
@vdemeester The test does *test* unmarshaling does not panic. 
Try running it w/o the changes to types_yaml.go!

@dnephin It does cast everything to string, instead of raising panics. I was thinking: the input file is text (readable and *writable* by humans) and the target datatype is string, so why not just use the string representation of the value. 

On the other hand, the orthodox docker-compose does issue an error, and ~~winter~~ #99 is coming. 
I'll update to return errors instead of throwing panics.

**Comment by imikushin** [comment]
Returning errors on incorrect datatypes now.

**Comment by imikushin** [comment]
@vdemeester @dnephin @ibuildthecloud I believe I've addressed the comments so far.

File: project/types_yaml.go [hunk 1]
```diff
@@ -80,7 +84,11 @@ func (s *Command) UnmarshalYAML(tag string, value interface{}) error {
 	case []interface{}:
 		parts := make([]string, len(value))
 		for k, v := range value {
-			parts[k] = v.(string)
+			if sv, ok := v.(string); ok {
+				parts[k] = sv
+			} else {
+				return fmt.Errorf("Cannot unmarshal '%v' of type %T into a string value", v, v)
+			}
```
Comment (dnephin) [inline project/types_yaml.go]: This block (including some of the existing lines) are repeated in quite a few places.  Could we make this a function?

```go
func ToStringSlice(value []interface{}) ([]string, error) {
    ...
}
```

Then many of these can be just:

```go
if s.parts, err = BuildParts(value); err != nil {
    return err
}
```

File: project/types_yaml.go [hunk 2]
```diff
@@ -226,13 +258,25 @@ func (s *MaporColonSlice) UnmarshalYAML(tag string, value interface{}) error {
...
-			parts = append(parts, strings.Join([]string{k.(string), v.(string)}, ":"))
+			if sk, ok := k.(string); ok {
+				if sv, ok := v.(string); ok {
+					parts = append(parts, sk+":"+sv)
+				} else {
+					return fmt.Errorf("Cannot unmarshal '%v' of type %T into a string value", v, v)
+				}
+			} else {
+				return fmt.Errorf("Cannot unmarshal '%v' of type %T into a string value", k, k)
+			}
```
Comment (dnephin) [inline project/types_yaml.go]: This is duplicated as well

**Comment by dnephin** [comment]
I'm not that familiar with this yaml parser. Why is it necessary that these values are cast to strings? Aren't they already strings? Is it trying to deal with cases where yaml treats them as numbers instead of strings?

**Comment by imikushin** [comment]
Parsed values can be anything: bool, integer, string. Consider this YAML list value, for example:
```yml
[1, "sick", list, of, false, assumptions]
```
Suppose, we have our own data type that uses `[]string` internally. When the parser is unmarshaling the given value and calling `UnmarshalYAML(tag string, value interface{})` method, here's what it's putting into the `value` argument:
```
[]interface{}{1, "sick", "list", "of", false, "assumptions"}
```
We've got an int, and a bool along with strings. So we have two options:
  - assert that each of the elements is a string, by using the type assertion `v.(string)`, and return an error otherwise (or throw a panic if we don't explicitly check the assertion is successful)
  - use a string representation of each value, by using `fmt.Sprint(v)`, and BTW this is what is used by the parser itself to unmarshal to `[]string` and what I've been proposing in the first place

File: project/types_yaml.go (hunk 1 above)
Comment (imikushin) [inline project/types_yaml.go]: Done.

File: project/types_yaml.go (hunk 2 above)
Comment (imikushin) [inline project/types_yaml.go]: Addressed.

**Comment by imikushin** [comment]
Done again.

**Comment by dnephin** [comment]
> assert that each of the elements is a string

I think this is what the strict yaml validation should be doing for us.

**Comment by imikushin** [comment]
> I think this is what the strict yaml validation should be doing for us.

Schema validation makes sure each element of the `[]interface{}` value is a `string`. 
Two things:
  - throwing panics is not nice when the API suggests an error can be returned
  - We don't have schema validation yet. When we have it, we'll just drop the `if !ok {...}` part which will become redundant. 

So, I propose checking for type casting errors (that's what this code is doing now) and drop these checks when the strict yaml validation is in. 
Alternatively, drop them now and face the panics OR coerce all values to strings via `fmt.Sprint(v)` (my original proposal). 

Pick one :game_die:

**Comment by imikushin** [comment]
@vdemeester Opinions welcome :smile:

**Comment by ibuildthecloud** [comment]
As I read this it seems that the below will blow up.

```
service:
  labels:
    foo: true
```

I believe we had this discussion before that while it would be nice if we didn't marshal types to string, the precedence has already been set in that docker-compose 1.5 and earlier will accept that form.  Until docker-compose changes to reject that form, we must support it.  I know personally that I have a lot of compose files with an unquoted `true` that will blow up.

**Comment by ibuildthecloud** [comment]
I'm trying to think what is the correct behavior here.  docker-compose doesn't actually blow up if you pass

```
labels:
  foo: true
```

The parser succeeds but then the Docker API itself will then blow up with some go marshaling API error.  So I kind of lean towards doing the same thing.  This change makes our parser more pedantic than docker-compose itself.  @dnephin WDYT?

**Comment by dnephin** [comment]
I think in 1.5.2, it should give a warning when a boolean is used. Previously we were only warning for booleans in the `environment` field.

I'm happy with more pedantic, as I think that's the direction that compose is going.

**Comment by ibuildthecloud** [comment]
@dnephin It's a hard pill to swallow for me that `foo: true` is not supported.  Is there docker-compose issue I can comment on or see more context why you guys have decided to go this route?  This just seems like a really bad user experience.

**Comment by ibuildthecloud** [comment]
@dnephin I've been reading through the yaml spec and understand now the complexities.

**Comment by ibuildthecloud** [comment]
LGTM

**Comment by imikushin** [comment]
@vdemeester Merge? :wink:

**Comment by vdemeester** [comment]
@imikushin yes :stuck_out_tongue_closed_eyes: 
LGTM 🐼

**Comment by imikushin** [comment]
Thx!

**Pull request merged**